fml/synchronization/sync_switch.h

```diff
@@ -0,0 +1,54 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FML_SYNCHRONIZATION_SYNC_SWITCH_H_
+#define FLUTTER_FML_SYNCHRONIZATION_SYNC_SWITCH_H_
+
+#include <forward_list>
+#include <functional>
+#include <mutex>
+#include "flutter/fml/macros.h"
+
+namespace fml {
+
+/// A threadsafe structure that allows you to switch between 2 different
+/// execution paths.
+class SyncSwitch {
+ public:
+  /// Represents the 2 code paths available when calling |SyncSwitch::Execute|.
+  struct Handlers {
+    /// Creates a |Handlers| were all exection paths are noops.
+    Handlers();
```

Comment:
Yep, if you ever want to call Execute you'll need one.

Reply:
`Handers(function on_success, function on_error)` and get rid of the setter in that case?

Reply:
Na, c++ doesn't have named parameters.  This pattern allows you to fake named parameters, check out the usage.  Since the 2 arguments are the same type it would be easy to get them mixed up and it's harder to read.